This neural network is designed to learn the relationship between two numbers (x & y) and predict the y value based on an x input.

The relationship for this model is known and in traditional programming can be expressed as:

In [1]:
def calculate_y(x):
    y = (2 * x) - 1
    return y

The first step is to import all required modules to build and run the neural network:
*   TensorFlow: TensorFlow is imported as tf for ease of use
*   keras: keras is the framework for defining a neural network as a set of Sequential layers
*   numpy: numpy helps represent the data as lists easily and quickly
*   time: Used to benchmark / time the training time of the model



In [2]:
import tensorflow as tf
from tensorflow import keras
#from tensorflow.keras import layers
import numpy as np
import time

Next, the neural network is defined. This is the simplest possible neural network. It has 1 layer and that layer has 1 neuron. The input shape to it is just 1 value.

In [3]:
model = keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

The next step is to compile the neural network. To do so, 2 functions need to be specified, the loss and the optimiser.

When the model is trying to learn the relationship between the two numbers, it makes a guess. The **loss** function measures the guessed answers against the known correct answers and determines how well or badly it did.

It then uses the **optimiser** function to make another guess. Based on what the loss function measured, it will try to minimise the loss.

The functions used are **Mean Squared Error** for the loss function and the **Stochastic Gradient Descent** for the optimiser function.

It will repeat this process for the number of **epochs** specified. An epoch is a term used in machine learning to indicate the number of passes of the entire training dataset the algorithm has completed.

In [4]:
model.compile(optimizer='sgd', loss='mean_squared_error')

Next, the dataset to train the model needs to be provided. The **Numpy** array type data structures are used by specifying the values as an np.array[ ]. 

In [5]:
xs = np.array([-1.0, 0.0, 1.0, 2.0, 3.0, 4.0,], dtype = float)
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0,], dtype = float)

The neural network is now ready to be trained (learn the relationship between x and y). 

The number of epochs used is specified as 1260 as trials established there is no accuracy improvement past this point for this model. Thus any further iterations would consume computing power and yield no results.

In [6]:
start_time = time.time()
model.fit(xs, ys, epochs = 1260)
delta = time.time() - start_time
print('Training time is {delta} seconds.'.format(delta = delta))

Epoch 1/1260
1/1 [==============================] - 0s 193ms/step - loss: 10.7090
Epoch 2/1260
1/1 [==============================] - 0s 999us/step - loss: 8.6406
Epoch 3/1260
1/1 [==============================] - 0s 997us/step - loss: 7.0088
Epoch 4/1260
1/1 [==============================] - 0s 998us/step - loss: 5.7207
Epoch 5/1260
1/1 [==============================] - 0s 997us/step - loss: 4.7030
Epoch 6/1260
1/1 [==============================] - 0s 998us/step - loss: 3.8982
Epoch 7/1260
1/1 [==============================] - 0s 0s/step - loss: 3.2609
Epoch 8/1260
1/1 [==============================] - 0s 999us/step - loss: 2.7556
Epoch 9/1260
1/1 [==============================] - 0s 998us/step - loss: 2.3541
Epoch 10/1260
1/1 [==============================] - 0s 998us/step - loss: 2.0344
Epoch 11/1260
1/1 [==============================] - 0s 0s/step - loss: 1.7791
Epoch 12/1260
1/1 [==============================] - 0s 997us/step - loss: 1.5746
Epoch 13/1260
1/1 [===========

1/1 [==============================] - 0s 0s/step - loss: 0.1402
Epoch 102/1260
1/1 [==============================] - 0s 998us/step - loss: 0.1373
Epoch 103/1260
1/1 [==============================] - 0s 997us/step - loss: 0.1345
Epoch 104/1260
1/1 [==============================] - 0s 998us/step - loss: 0.1317
Epoch 105/1260
1/1 [==============================] - 0s 0s/step - loss: 0.1290
Epoch 106/1260
1/1 [==============================] - 0s 999us/step - loss: 0.1264
Epoch 107/1260
1/1 [==============================] - 0s 998us/step - loss: 0.1238
Epoch 108/1260
1/1 [==============================] - 0s 996us/step - loss: 0.1212
Epoch 109/1260
1/1 [==============================] - 0s 998us/step - loss: 0.1187
Epoch 110/1260
1/1 [==============================] - 0s 999us/step - loss: 0.1163
Epoch 111/1260
1/1 [==============================] - 0s 998us/step - loss: 0.1139
Epoch 112/1260
1/1 [==============================] - 0s 995us/step - loss: 0.1116
Epoch 113/1260
1/1 [=====

1/1 [==============================] - 0s 0s/step - loss: 0.0180
Epoch 201/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0176
Epoch 202/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0172
Epoch 203/1260
1/1 [==============================] - 0s 0s/step - loss: 0.0169
Epoch 204/1260
1/1 [==============================] - 0s 997us/step - loss: 0.0165
Epoch 205/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0162
Epoch 206/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0159
Epoch 207/1260
1/1 [==============================] - 0s 0s/step - loss: 0.0155
Epoch 208/1260
1/1 [==============================] - 0s 999us/step - loss: 0.0152
Epoch 209/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0149
Epoch 210/1260
1/1 [==============================] - 0s 996us/step - loss: 0.0146
Epoch 211/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0143
Epoch 212/1260
1/1 [========

1/1 [==============================] - 0s 998us/step - loss: 0.0023
Epoch 300/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0023
Epoch 301/1260
1/1 [==============================] - 0s 997us/step - loss: 0.0022
Epoch 302/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0022
Epoch 303/1260
1/1 [==============================] - 0s 0s/step - loss: 0.0021
Epoch 304/1260
1/1 [==============================] - 0s 999us/step - loss: 0.0021
Epoch 305/1260
1/1 [==============================] - 0s 997us/step - loss: 0.0020
Epoch 306/1260
1/1 [==============================] - 0s 999us/step - loss: 0.0020
Epoch 307/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0019
Epoch 308/1260
1/1 [==============================] - 0s 997us/step - loss: 0.0019
Epoch 309/1260
1/1 [==============================] - 0s 998us/step - loss: 0.0019
Epoch 310/1260
1/1 [==============================] - 0s 997us/step - loss: 0.0018
Epoch 311/1260
1/1 [==

1/1 [==============================] - 0s 998us/step - loss: 3.1381e-04
Epoch 396/1260
1/1 [==============================] - 0s 0s/step - loss: 3.0737e-04
Epoch 397/1260
1/1 [==============================] - 0s 997us/step - loss: 3.0105e-04
Epoch 398/1260
1/1 [==============================] - 0s 998us/step - loss: 2.9487e-04
Epoch 399/1260
1/1 [==============================] - 0s 997us/step - loss: 2.8881e-04
Epoch 400/1260
1/1 [==============================] - 0s 998us/step - loss: 2.8288e-04
Epoch 401/1260
1/1 [==============================] - 0s 998us/step - loss: 2.7707e-04
Epoch 402/1260
1/1 [==============================] - 0s 996us/step - loss: 2.7138e-04
Epoch 403/1260
1/1 [==============================] - 0s 0s/step - loss: 2.6580e-04
Epoch 404/1260
1/1 [==============================] - 0s 998us/step - loss: 2.6034e-04
Epoch 405/1260
1/1 [==============================] - 0s 998us/step - loss: 2.5500e-04
Epoch 406/1260
1/1 [==============================] - 0s 998us/s

Epoch 490/1260
1/1 [==============================] - 0s 998us/step - loss: 4.3688e-05
Epoch 491/1260
1/1 [==============================] - 0s 0s/step - loss: 4.2791e-05
Epoch 492/1260
1/1 [==============================] - 0s 998us/step - loss: 4.1911e-05
Epoch 493/1260
1/1 [==============================] - 0s 998us/step - loss: 4.1051e-05
Epoch 494/1260
1/1 [==============================] - 0s 998us/step - loss: 4.0208e-05
Epoch 495/1260
1/1 [==============================] - 0s 998us/step - loss: 3.9381e-05
Epoch 496/1260
1/1 [==============================] - 0s 998us/step - loss: 3.8573e-05
Epoch 497/1260
1/1 [==============================] - 0s 998us/step - loss: 3.7780e-05
Epoch 498/1260
1/1 [==============================] - 0s 0s/step - loss: 3.7005e-05
Epoch 499/1260
1/1 [==============================] - 0s 998us/step - loss: 3.6244e-05
Epoch 500/1260
1/1 [==============================] - 0s 997us/step - loss: 3.5501e-05
Epoch 501/1260
1/1 [=============================

1/1 [==============================] - 0s 997us/step - loss: 6.2097e-06
Epoch 585/1260
1/1 [==============================] - 0s 998us/step - loss: 6.0819e-06
Epoch 586/1260
1/1 [==============================] - 0s 997us/step - loss: 5.9573e-06
Epoch 587/1260
1/1 [==============================] - ETA: 0s - loss: 5.8348e-0 - 0s 998us/step - loss: 5.8348e-06
Epoch 588/1260
1/1 [==============================] - 0s 998us/step - loss: 5.7146e-06
Epoch 589/1260
1/1 [==============================] - 0s 997us/step - loss: 5.5975e-06
Epoch 590/1260
1/1 [==============================] - 0s 997us/step - loss: 5.4822e-06
Epoch 591/1260
1/1 [==============================] - 0s 0s/step - loss: 5.3700e-06
Epoch 592/1260
1/1 [==============================] - 0s 997us/step - loss: 5.2595e-06
Epoch 593/1260
1/1 [==============================] - 0s 998us/step - loss: 5.1515e-06
Epoch 594/1260
1/1 [==============================] - 0s 998us/step - loss: 5.0453e-06
Epoch 595/1260
1/1 [=============

1/1 [==============================] - 0s 997us/step - loss: 8.8271e-07
Epoch 679/1260
1/1 [==============================] - 0s 998us/step - loss: 8.6447e-07
Epoch 680/1260
1/1 [==============================] - 0s 998us/step - loss: 8.4681e-07
Epoch 681/1260
1/1 [==============================] - 0s 0s/step - loss: 8.2938e-07
Epoch 682/1260
1/1 [==============================] - 0s 999us/step - loss: 8.1234e-07
Epoch 683/1260
1/1 [==============================] - 0s 996us/step - loss: 7.9571e-07
Epoch 684/1260
1/1 [==============================] - 0s 0s/step - loss: 7.7931e-07
Epoch 685/1260
1/1 [==============================] - 0s 998us/step - loss: 7.6330e-07
Epoch 686/1260
1/1 [==============================] - 0s 997us/step - loss: 7.4767e-07
Epoch 687/1260
1/1 [==============================] - 0s 997us/step - loss: 7.3246e-07
Epoch 688/1260
1/1 [==============================] - 0s 997us/step - loss: 7.1740e-07
Epoch 689/1260
1/1 [==============================] - 0s 998us/s

1/1 [==============================] - 0s 998us/step - loss: 1.2557e-07
Epoch 773/1260
1/1 [==============================] - 0s 0s/step - loss: 1.2303e-07
Epoch 774/1260
1/1 [==============================] - 0s 997us/step - loss: 1.2050e-07
Epoch 775/1260
1/1 [==============================] - 0s 996us/step - loss: 1.1798e-07
Epoch 776/1260
1/1 [==============================] - 0s 997us/step - loss: 1.1556e-07
Epoch 777/1260
1/1 [==============================] - 0s 998us/step - loss: 1.1316e-07
Epoch 778/1260
1/1 [==============================] - 0s 998us/step - loss: 1.1088e-07
Epoch 779/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0861e-07
Epoch 780/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0638e-07
Epoch 781/1260
1/1 [==============================] - 0s 999us/step - loss: 1.0417e-07
Epoch 782/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0208e-07
Epoch 783/1260
1/1 [==============================] - 0s 997u

1/1 [==============================] - 0s 997us/step - loss: 1.7908e-08
Epoch 867/1260
1/1 [==============================] - 0s 998us/step - loss: 1.7513e-08
Epoch 868/1260
1/1 [==============================] - 0s 998us/step - loss: 1.7161e-08
Epoch 869/1260
1/1 [==============================] - 0s 998us/step - loss: 1.6807e-08
Epoch 870/1260
1/1 [==============================] - 0s 996us/step - loss: 1.6462e-08
Epoch 871/1260
1/1 [==============================] - 0s 998us/step - loss: 1.6139e-08
Epoch 872/1260
1/1 [==============================] - 0s 999us/step - loss: 1.5804e-08
Epoch 873/1260
1/1 [==============================] - 0s 997us/step - loss: 1.5466e-08
Epoch 874/1260
1/1 [==============================] - 0s 998us/step - loss: 1.5138e-08
Epoch 875/1260
1/1 [==============================] - 0s 997us/step - loss: 1.4843e-08
Epoch 876/1260
1/1 [==============================] - 0s 0s/step - loss: 1.4537e-08
Epoch 877/1260
1/1 [==============================] - 0s 998u

1/1 [==============================] - 0s 998us/step - loss: 2.5775e-09
Epoch 961/1260
1/1 [==============================] - 0s 998us/step - loss: 2.5226e-09
Epoch 962/1260
1/1 [==============================] - 0s 998us/step - loss: 2.4717e-09
Epoch 963/1260
1/1 [==============================] - 0s 997us/step - loss: 2.4208e-09
Epoch 964/1260
1/1 [==============================] - 0s 998us/step - loss: 2.3677e-09
Epoch 965/1260
1/1 [==============================] - 0s 998us/step - loss: 2.3188e-09
Epoch 966/1260
1/1 [==============================] - 0s 997us/step - loss: 2.2777e-09
Epoch 967/1260
1/1 [==============================] - 0s 996us/step - loss: 2.2260e-09
Epoch 968/1260
1/1 [==============================] - 0s 996us/step - loss: 2.1753e-09
Epoch 969/1260
1/1 [==============================] - 0s 0s/step - loss: 2.1357e-09
Epoch 970/1260
1/1 [==============================] - 0s 998us/step - loss: 2.0891e-09
Epoch 971/1260
1/1 [==============================] - 0s 998u

1/1 [==============================] - 0s 996us/step - loss: 3.6999e-10
Epoch 1055/1260
1/1 [==============================] - 0s 0s/step - loss: 3.6411e-10
Epoch 1056/1260
1/1 [==============================] - 0s 998us/step - loss: 3.5597e-10
Epoch 1057/1260
1/1 [==============================] - 0s 998us/step - loss: 3.4729e-10
Epoch 1058/1260
1/1 [==============================] - 0s 998us/step - loss: 3.4409e-10
Epoch 1059/1260
1/1 [==============================] - 0s 997us/step - loss: 3.3553e-10
Epoch 1060/1260
1/1 [==============================] - 0s 998us/step - loss: 3.2988e-10
Epoch 1061/1260
1/1 [==============================] - 0s 998us/step - loss: 3.2398e-10
Epoch 1062/1260
1/1 [==============================] - 0s 0s/step - loss: 3.1580e-10
Epoch 1063/1260
1/1 [==============================] - 0s 998us/step - loss: 3.1019e-10
Epoch 1064/1260
1/1 [==============================] - 0s 997us/step - loss: 3.0414e-10
Epoch 1065/1260
1/1 [==============================] -

Epoch 1148/1260
1/1 [==============================] - 0s 0s/step - loss: 5.3646e-11
Epoch 1149/1260
1/1 [==============================] - 0s 998us/step - loss: 5.3016e-11
Epoch 1150/1260
1/1 [==============================] - 0s 998us/step - loss: 5.2395e-11
Epoch 1151/1260
1/1 [==============================] - 0s 0s/step - loss: 5.0666e-11
Epoch 1152/1260
1/1 [==============================] - 0s 997us/step - loss: 4.9592e-11
Epoch 1153/1260
1/1 [==============================] - 0s 999us/step - loss: 4.9409e-11
Epoch 1154/1260
1/1 [==============================] - 0s 998us/step - loss: 4.9009e-11
Epoch 1155/1260
1/1 [==============================] - 0s 998us/step - loss: 4.7502e-11
Epoch 1156/1260
1/1 [==============================] - 0s 997us/step - loss: 4.7323e-11
Epoch 1157/1260
1/1 [==============================] - 0s 998us/step - loss: 4.5953e-11
Epoch 1158/1260
1/1 [==============================] - 0s 0s/step - loss: 4.5328e-11
Epoch 1159/1260
1/1 [====================

1/1 [==============================] - 0s 997us/step - loss: 1.0604e-11
Epoch 1242/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0604e-11
Epoch 1243/1260
1/1 [==============================] - 0s 997us/step - loss: 1.0604e-11
Epoch 1244/1260
1/1 [==============================] - 0s 997us/step - loss: 1.0604e-11
Epoch 1245/1260
1/1 [==============================] - 0s 0s/step - loss: 1.0604e-11
Epoch 1246/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0604e-11
Epoch 1247/1260
1/1 [==============================] - 0s 997us/step - loss: 1.0604e-11
Epoch 1248/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0604e-11
Epoch 1249/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0604e-11
Epoch 1250/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0604e-11
Epoch 1251/1260
1/1 [==============================] - 0s 998us/step - loss: 1.0604e-11
Epoch 1252/1260
1/1 [==============================

Finally, now that the model has been trained, prediction for inputs with unknown results can be made. An input of x = 10 has been chosen at random.

In [7]:
print(model.predict([10.0]))

[[18.999987]]


The result is not precisely what was expected. A value of 19 was expected but a value of 18.999987 was returned. This is due to the fact that the model deals with probabilities and not certainties and the limited sample size of 6 value pairs does not allow the model to calculate the relationship to the exact precise number.